urler.1
.\" You can view this file with:
.\" man -l urler.1
.\" Written by Daniel Stenberg
.\"
.TH urler 1 "31 Mar 2023" "urler 0.1" "urler Manual"
.SH NAME
urler \- parse and manipulate URLs
.SH SYNOPSIS
.B urler [options]
.SH DESCRIPTION
.B urler
parses, manipulates and outputs URLs and parts of URLs.

It uses the RFC 3986 definition of URLs and it uses libcurl's URL parser to do
so, which includes a few "extensions".

Typically you set a few inputs and decide what of that you want output.
.SH "MISC OPTIONS"
.IP "-h, --help"
Show the help output.
.IP "-v, --version"
Show version information and exit.
.SH "INPUT OPTIONS"
.IP "--url [URL]"
Set the input URL to work with. The URL may be provided without a scheme,
which then typically is not actually a legal URL but urler will try to figure
out what is meant and guess what scheme to use.

Providing multiple URLs will make urler act on all URLs in a serial fashion.
.IP "--url-file [name]"
Read URLs to work on from the given file. Use the file name "-" (a single
minus) to tell urler to read the URLs from stdin.

Each line needs to be a single valid URL and the line needs to end with a
newline. The maximum URL length supported in a file like this is 4095 bytes.
.IP "--redirect [URL]"
Redirect the base URL to this. It requires that you set the base url with \fB--url\fP
.IP "--append [component]=[data]"
Append data to a component. This can only append data to the path and the
query component.

For path:, this URL encodes and appends this segment to the path, separated
with a slash.

For query, this URL encodes and appends this segment to the query, separated
with an ampersand (&). If the segment contains an equal sign ('=') that will
be kept verbatim and both sides of the first occrance will be URL encoded
separately.
.IP "--set [component]=[data]"
Set this URL component. Setting blank string ("") will clear it from the
URL.

The following components can be set: url, scheme, user, password,
options, host, port, path, query, fragment and zoneid.
.SH "OUTPUT OPTIONS"
.IP "--get [format]"
Output data according to the provided format string. Components from the URL
can be output when specified as \fB{component}\fP, with the name of the part
within curly braces. The following component names are available: {url},
{scheme}, {user}, {password}, {options}, {host}, {port}, {path}, {query},
{fragment} and {zoneid}.
.SH "MODIFIERS"
.IP "--urldecode"
By default each component is output using the format of the input (URL
encoded). This option makes urler instead provide it URL decoded. Note that
%20 will then output space and %09 a tab etc.
.SH EXAMPLES
.IP "Replace the host name of a URL"
.nf
$ urler --url https://curl.se --set host=example.com
https://example.com/
.fi
.IP "Create a URL by setting components"
.nf
 $ urler --set host=example.com --set scheme=ftp
 ftp://example.com/
.fi
.IP "Redirect a URL"
.nf
$ urler --url https://curl.se/we/are.html --redirect here.html
https://curl.se/we/here.html
.fi
.IP "Change port number"
This also shows how urler will remove dot-dot sequences

.nf
$ urler --url https://curl.se/we/../are.html --set port=8080
https://curl.se:8080/are.html
.IP "Extract the path from a URL"
.nf
$ urler --url https://curl.se/we/are.html --get '{path}'
/we/are.html
.IP "Extract the port from a URL"
This gets the default port based on the scheme if the port is not set in the
URL.

.nf
$ urler --url https://curl.se/we/are.html --get '{port}'
443
.IP "Append a path segment to a URL"
.nf
$ urler --url https://curl.se/hello --append path=you
https://curl.se/hello/you
.IP "Append a query segment to a URL"
.nf
$ urler --url "https://curl.se?name=hello" --append query=search=string
 https://curl.se/?name=hello&search=string
.IP "Read URLs from stdin"
.nf
$ cat urllist.txt | urler --url-file -
...
.SH WWW
https://github.com/curl/urler


RELEASE-NOTES
trurl 0.3

Changes since previous release

 * the set port is the same as the scheme-default, don't show
 * many more tests
 * https://curl.se/trurl is now the project's website
 * added --accept-space
 * added --trim

Bugfixes since previous release

 * setting port to a blank now clears it
 * improved the man page
 * fix --append query when no previous query exists
 * output JSON as an array
 * provide the -h list in pure alphabetical order
 * add an INSTALL variable to the Makefile
 * enable building with msvc on Windows
 * use 0644 for the man page install
 * use Test::More inside of test.pl
 * nicer JSON tests and test output


urler.c
/***************************************************************************
 *                                  _   _ ____  _
 *  Project                     ___| | | |  _ \| |
 *                             / __| | | | |_) | |
 *                            | (__| |_| |  _ <| |___
 *                             \___|\___/|_| \_\_____|
 *
 * Daniel Stenberg, <daniel@haxx.se>, et al.
 *
 * This software is licensed as described in the file COPYING, which
 * you should have received as part of this distribution. The terms
 * are also available at https://curl.se/docs/copyright.html.
 *
 * You may opt to use, copy, modify, merge, publish, distribute and/or sell
 * copies of the Software, and permit persons to whom the Software is
 * furnished to do so, under the terms of the COPYING file.
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
 * KIND, either express or implied.
 *
 * SPDX-License-Identifier: curl
 *
 ***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <curl/curl.h>
#include <curl/mprintf.h>

#include "version.h"

#define OUTPUT_URL      0  /* default */
#define OUTPUT_SCHEME   1
#define OUTPUT_USER     2
#define OUTPUT_PASSWORD 3
#define OUTPUT_OPTIONS  4
#define OUTPUT_HOST     5
#define OUTPUT_PORT     6
#define OUTPUT_PATH     7
#define OUTPUT_QUERY    8
#define OUTPUT_FRAGMENT 9
#define OUTPUT_ZONEID   10
#define PROGNAME        "urler"

static void help(const char *msg)
{
  if(msg != NULL)
    fprintf(stderr, "%s:\n\n", msg);
  fprintf(stderr, "Usage: [options] [URL]\n"
          "  -h,--help                   - this help\n"
          "  -v,--version                - show version\n"
          " INPUT\n"
          "  --append [component]=[data] - append data to component\n"
          "  --redirect [URL]            - redirect the base URL to this\n"
          "  --set [component]=[data]    - set this component\n"
          "  --url [base URL]            - URL to start with\n"
          "  --url-file [file/-]         - read URLs from file or stdin\n"
          " OUTPUT\n"
          "  --get [format]              - output URL components\n"
          " MODIFIERS\n"
          "  --urldecode                 - URL decode the output\n"
    );
  exit(1);
}

static void show_version(void)
{
  curl_version_info_data *data = curl_version_info(CURLVERSION_NOW);
  fprintf(stdout, "%s version %s libcurl/%s\n", PROGNAME, URLER_VERSION_TXT, data->version);
  exit(0);
}

struct option {
  struct curl_slist *url_list;
  struct curl_slist *append_path;
  struct curl_slist *append_query;
  struct curl_slist *set_list;
  const char *redirect;
  const char *format;

  unsigned int urldecode:1;
  unsigned char output;
};

static void urladd(struct option *o, const char *url)
{
  struct curl_slist *n;
  n = curl_slist_append(o->url_list, url);
  if(n)
    o->url_list = n;
}


/* read URLs from this file/stdin */
static void urlfile(struct option *o, const char *file)
{
  FILE *f;
  bool closeit = false;

  if(strcmp("-", file)) {
    f = fopen(file, "rt");
    if(!f)
      help("--url-file not found");
    closeit = true;
  }
  else
    f = stdin;
  if(f) {
    char buffer[4096]; /* arbitrary max */
    while(fgets(buffer, sizeof(buffer), f)) {
      char *eol = strchr(buffer, '\n');
      if(eol && (eol > buffer)) {
        if(eol[-1] == '\r')
          /* CRLF detected */
          eol--;
        *eol = 0; /* end of URL */
        urladd(o, buffer);
      }
      else {
        /* no newline or no content, skip */
      }
    }
    if(closeit)
      fclose(f);
  }
}

static void pathadd(struct option *o, const char *path)
{
  struct curl_slist *n;
  char *urle = curl_easy_escape(NULL, path, 0);
  if(urle) {
    n = curl_slist_append(o->append_path, urle);
    if(n) {
      o->append_path = n;
    }
    free(urle);
  }
}

static void queryadd(struct option *o, const char *query)
{
  struct curl_slist *n;
  char *p = strchr(query, '=');
  char *urle;
  if(p) {
    /* URL encode the left and the right side of the '=' separately */
    char *f1 = curl_easy_escape(NULL, query, p - query);
    char *f2 = curl_easy_escape(NULL, p + 1, 0);
    urle = curl_maprintf("%s=%s", f1, f2);
    curl_free(f1);
    curl_free(f2);
  }
  else
    urle = curl_easy_escape(NULL, query, 0);
  if(urle) {
    n = curl_slist_append(o->append_query, urle);
    if(n) {
      o->append_query = n;
    }
    free(urle);
  }
}

struct var {
  const char *name;
  CURLUPart part;
};

static const struct var variables[] = {
  {"url",      CURLUPART_URL},
  {"scheme",   CURLUPART_SCHEME},
  {"user",     CURLUPART_USER},
  {"password", CURLUPART_PASSWORD},
  {"options",  CURLUPART_OPTIONS},
  {"host",     CURLUPART_HOST},
  {"port",     CURLUPART_PORT},
  {"path",     CURLUPART_PATH},
  {"query",    CURLUPART_QUERY},
  {"fragment", CURLUPART_FRAGMENT},
  {"zoneid",   CURLUPART_ZONEID},
  {NULL, 0}
};


static void appendadd(struct option *o,
                      const char *arg)
{
  if(!strncasecmp("path=", arg, 5))
    pathadd(o, arg + 5);
  else if(!strncasecmp("query=", arg, 6))
    queryadd(o, arg + 6);
  else
    help("--append unsupported component");
}

static void setadd(struct option *o,
              const char *set) /* [component]=[data] */
{
  struct curl_slist *n;
  n = curl_slist_append(o->set_list, set);
  if(n)
    o->set_list = n;
}

static bool checkoptarg(const char *str,
                   const char *given,
                   const char *arg)
{
  if(!strcmp(str, given)) {
    if(!arg) {
      char buffer[128];
      curl_msnprintf(buffer, sizeof(buffer),
                     "Missing argument for %s", str);
      help(buffer);
    }
    return true;
  }
  return false;
}

static int getlongarg(struct option *op,
                      const char *flag,
                      const char *arg,
                      int *usedarg)
{
  *usedarg = 0;
  if(!strcmp("--help", flag))
    help(NULL);
  else if(!strcmp("--version", flag))
    show_version();
  else if(checkoptarg("--url", flag, arg)) {
    urladd(op, arg);
    *usedarg = 1;
  }
  else if(checkoptarg("--url-file", flag, arg)) {
    urlfile(op, arg);
    *usedarg = 1;
  }
  else if(checkoptarg("--append", flag, arg)) {
    appendadd(op, arg);
    *usedarg = 1;
  }
  else if(checkoptarg("--set", flag, arg)) {
    setadd(op, arg);
    *usedarg = 1;
  }
  else if(checkoptarg("--redirect", flag, arg)) {
    op->redirect = arg;
    *usedarg = 1;
  }
  else if(checkoptarg("--get", flag, arg)) {
    op->format = arg;
    *usedarg = 1;
  }
  else if(!strcmp("--urldecode", flag))
    op->urldecode = 1;
  else
    return 1;  /* unrecognized option */
  return 0;
}

static int getshortarg(struct option *op,
                       const char *flag)
{
  (void)op;
  if(!strcmp("-v", flag))
    show_version();
  else if(!strcmp("-h", flag))
    help(NULL);
  else
    return 1;  /* unrecognized option */
  return 0;
}

static void format(struct option *op, CURLU *uh)
{
  FILE *stream = stdout;
  const char *ptr = op->format;
  bool done = false;

  while(ptr && *ptr && !done) {
    if('{' == *ptr) {
      if('{' == ptr[1]) {
        /* an escaped {-letter */
        fputc('{', stream);
        ptr += 2;
      }
      else {
        /* this is meant as a variable to output */
        char *end;
        size_t vlen;
        int i;
        end = strchr(ptr, '}');
        ptr++; /* pass the { */
        if(!end) {
          /* syntax error */
          continue;
        }
        vlen = end - ptr;
        for(i = 0; variables[i].name; i++) {
          if((strlen(variables[i].name) == vlen) &&
             !strncasecmp(ptr, variables[i].name, vlen)) {
            char *nurl;
            if(!curl_url_get(uh, variables[i].part, &nurl, CURLU_DEFAULT_PORT|
                             (op->urldecode?CURLU_URLDECODE:0))) {
              fprintf(stream, "%s", nurl);
              curl_free(nurl);
            }
            break;
          }
        }
        ptr = end + 1; /* pass the end */
      }
    }
    else if('\\' == *ptr && ptr[1]) {
      switch(ptr[1]) {
      case 'r':
        fputc('\r', stream);
        break;
      case 'n':
        fputc('\n', stream);
        break;
      case 't':
        fputc('\t', stream);
        break;
      default:
        /* unknown, just output this */
        fputc(*ptr, stream);
        fputc(ptr[1], stream);
        break;
      }
      ptr += 2;
    }
    else {
      fputc(*ptr, stream);
      ptr++;
    }
  }
  fputc('\n', stream);
}

static void set(CURLU *uh,
                struct option *o)
{
  struct curl_slist *node;
  for(node =  o->set_list; node; node=node->next) {
    char *set = node->data;
    int i;
    char *ptr = strchr(set, '=');
    if(ptr) {
      size_t vlen = ptr-set;
      bool found = false;
      for(i=0; variables[i].name; i++) {
        if((strlen(variables[i].name) == vlen) &&
           !strncasecmp(set, variables[i].name, vlen)) {
          curl_url_set(uh, variables[i].part, ptr+1, CURLU_NON_SUPPORT_SCHEME);
          found = true;
          break;
        }
      }
      if(!found)
        help("Set unknown component");
    }
  }
}

int main(int argc, const char **argv)
{
  int exit_status = 0;
  char *nurl = NULL;
  struct option o;
  CURLU *uh;
  struct curl_slist *node;
  memset(&o, 0, sizeof(o));
  curl_global_init(CURL_GLOBAL_ALL);

  for(argc--, argv++; argc > 0; argc--, argv++) {
    if(argv[0][0] == '-' && argv[0][1] != '-') {
      /* single-dash prefix */
      if(getshortarg(&o, argv[0]))
        help("unknown option");
    }
    else if(argv[0][0] == '-' && argv[0][1] == '-') {
      /* dash-dash prefixed */
      int usedarg = 0;
      if(getlongarg(&o, argv[0], argv[1], &usedarg))
        help("unknown option");

      if(usedarg) {
        /* skip the parsed argument */
        argc--;
        argv++;
      }
    }
    else {
      /* this is a URL */
      urladd(&o, argv[0]);
    }
  }

  node = o.url_list;
  do {
    const char *url = NULL;
    struct curl_slist *p;
    uh = curl_url();
    if(!uh)
      help("out of memory");
    if(node) {
      url = node->data;
      curl_url_set(uh, CURLUPART_URL, url,
                   CURLU_GUESS_SCHEME|CURLU_NON_SUPPORT_SCHEME);
      if(o.redirect)
        curl_url_set(uh, CURLUPART_URL, o.redirect,
                     CURLU_GUESS_SCHEME|CURLU_NON_SUPPORT_SCHEME);
    }
    /* set everything */
    set(uh, &o);

    /* append path segments */
    for(p = o.append_path; p; p=p->next) {
      char *apath = p->data;
      char *opath;
      char *npath;
      size_t olen;
      /* extract the current path */
      curl_url_get(uh, CURLUPART_PATH, &opath, 0);

      /* does the existing path end with a slash, then don't
         add one inbetween */
      olen = strlen(opath);

      /* append the new segment */
      npath = curl_maprintf("%s%s%s", opath,
                            opath[olen-1] == '/' ? "" : "/",
                            apath);
      if(npath) {
        /* set the new path */
        curl_url_set(uh, CURLUPART_PATH, npath, 0);
      }
      curl_free(npath);
      curl_free(opath);
    }

    /* append query segments */
    for(p = o.append_query; p; p=p->next) {
      char *aq = p->data;
      char *oq;
      char *nq;
      /* extract the current query */
      curl_url_get(uh, CURLUPART_QUERY, &oq, 0);

      /* append the new segment */
      nq = curl_maprintf("%s&%s", oq, aq);
      if(nq) {
        /* set the new query */
        curl_url_set(uh, CURLUPART_QUERY, nq, 0);
      }
      curl_free(nq);
      curl_free(oq);
    }

    if(o.format) {
      /* custom output format */
      format(&o, uh);
    }
    else {
      /* default output is full URL */
      if(!curl_url_get(uh, CURLUPART_URL, &nurl,
                       o.urldecode?CURLU_URLDECODE:0)) {
        printf("%s\n", nurl);
        curl_free(nurl);
      }
      else {
        fprintf(stderr, "not enough input for a URL (%s -h for help)\n",
                PROGNAME);
        exit_status = 1;
      }
    }
    curl_url_cleanup(uh);
    if(node)
      node = node->next;
  } while(node);
  /* we're done with libcurl, so clean it up */
  curl_slist_free_all(o.url_list);
  curl_slist_free_all(o.set_list);
  curl_slist_free_all(o.append_path);
  curl_slist_free_all(o.append_query);
  curl_global_cleanup();
  return exit_status;
}
